src/engine/Service.cpp

```diff
@@ -493,3 +488,33 @@ std::optional<std::string> Service::idToValueForValuesClause(
       }
   }
 }
+
+// ____________________________________________________________________________
+void Service::precomputeSiblingResult(std::shared_ptr<Operation> left,
+                                      std::shared_ptr<Operation> right,
+                                      bool rightOnly) {
+  AD_CORRECTNESS_CHECK(left && right);
+  auto a = std::dynamic_pointer_cast<Service>(left);
+  auto b = std::dynamic_pointer_cast<Service>(right);
+
+  if ((rightOnly && !b) ||
+      (!rightOnly && static_cast<bool>(a) == static_cast<bool>(b))) {
+    return;
+  }
+
+  const auto& [service, sibling] = [&]() {
+    if (a) {
+      return std::tie(a, right);
+    } else {
+      AD_CORRECTNESS_CHECK(b);
+      return std::tie(b, left);
+    }
+  }();
+
+  auto siblingResult = sibling->getResult();
+  sibling->precomputedResultBecauseSiblingOfService_ = siblingResult;
+  if (siblingResult->idTable().size() <=
+      RuntimeParameters().get<"service-max-value-rows">()) {
+    service->precomputedSiblingResult_ = siblingResult;
+  }
+}
```

Comment:
Although currently neither of the JOIN-like operations support laziness, we still need to prepare the support here.
This function should take another parameter `requestLaziness` which is then passed to the `getResult()` call.
If the result is in fact lazy (akak not `fullyMaterialized`, then you basically store the returned IdTables until either you've consumed everything, or you have too many rows. Then you create a new result with a new generator that first yields all the blocks you've already retrieved and then continues consuming the downstream generator.
For the service result you can just concatenate the tables to a single IdTable to get a materialized result.